### PR TITLE
Incrementally introduce TypeScript language

### PR DESCRIPTION
**Pull request by saramarcondes**
### Testing instructions

Checkout the branch and run `tsc --build` and ensure there are no errors. Manually check `utils/types` and ensure that they're being built correctly and as expected (especially anything re-exporting lodash).

For some reason this breaks the pre-commit hook in this manner:

```
📦  Zero test...

Configuration error:

Could not locate module @wp-g2/substate/types mapped as:
/Users/sara/projects/g2/packages/substate/types/src.

Please check your configuration for these entries:
{
  "moduleNameMapper": {
    "/^@wp-g2\/(.*)$/": "/Users/sara/projects/g2/packages/$1/src"
  },
  "resolver": null
}
    at createNoMappedModuleFoundError (/Users/sara/projects/g2/node_modules/jest-resolve/build/index.js:501:17)
    at Resolver.resolveStubModuleName (/Users/sara/projects/g2/node_modules/jest-resolve/build/index.js:462:21)
    at Resolver.getMockModule (/Users/sara/projects/g2/node_modules/jest-resolve/build/index.js:304:31)
    at /Users/sara/projects/g2/node_modules/jest-resolve-dependencies/build/index.js:67:45
    at Array.reduce (<anonymous>)
    at DependencyResolver.resolve (/Users/sara/projects/g2/node_modules/jest-resolve-dependencies/build/index.js:53:25)
    at DependencyResolver.resolveInverseModuleMap (/Users/sara/projects/g2/node_modules/jest-resolve-dependencies/build/index.js:175:30)
    at DependencyResolver.resolveInverse (/Users/sara/projects/g2/node_modules/jest-resolve-dependencies/build/index.js:198:17)
    at SearchSource.findRelatedTests (/Users/sara/projects/g2/node_modules/@jest/core/build/SearchSource.js:280:30)
    at SearchSource.findRelatedTestsFromPattern (/Users/sara/projects/g2/node_modules/@jest/core/build/SearchSource.js:345:19)
husky > pre-commit hook failed (add --no-verify to bypass)
```

### CODE REVIEW

**Comment by ItsJonQ** [comment]
> For some reason this breaks the pre-commit hook in this manner:

Oh no... built tool stuff 🙈 

Is it because the `types/index.d.ts` file is deleted?

---

That aside.. I'm not sure how to review this PR 😊

**Comment by saramarcondes** [comment]
> Is it because the types/index.d.ts file is deleted?

It's deleted from git because now it's being auto-generated. Locally the file does exist due to the build 🤔 Why does jest care about the test files though?

File: tsconfig.base.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+	"compilerOptions": {
+		"allowJs": true,
```
Comment (nicolad) [inline tsconfig.base.json:3]: :heart:

**Comment by ItsJonQ** [comment]
@saramarcondes Oh wow! Look at them types 😍 

I'm not too sure the pre-commit hook / jest would be made at your changes 🙈 .

Maybe it's okay from this point forward? (I'm guessing you committed the stuff with `--no-verify`)

**Review by sirreal — APPROVED** [review]
I rebased to fix a conflict. Also:

- Improve `types` directory ignores
- Exclude non-generated `types` directories from ignore
- Add type clean to clean script

Let's land this and see how it goes, thanks everyone!

**Comment by saramarcondes** [comment]
> Maybe it's okay from this point forward?

It doesn't seem to be. Every time I've tried to change a JS file it breaks on the test run from pre-commit. @sirreal did you notice this happening for you? I don't think this is safe to land if it's breaking the pre-commit hook.

**Comment by sirreal** [comment]
> @sirreal did you notice this happening for you?

I had noticed this, but it didn't happen during my rebase. I'd prefer to push forward until it becomes a problem, if it's only an issue on certain commits we're probably fine. The end goal is for this project to be integrated in Gutenberg and the tooling that's having issues won't be used so I'd like to avoid investing time there.

**Comment by ItsJonQ** [comment]
> The end goal is for this project to be integrated in Gutenberg and the tooling that's having issues won't be used so I'd like to avoid investing time there

I agree 👍 . I did ma best with [@itsjonq/zero](https://github.com/itsjonq/zero). This was the biggest project my scripts package has ever needed to manage 😊

**Pull request merged**